yb-voyager/src/anon/sql_anonymizer.go

```diff
@@ -396,111 +480,794 @@ func (a *SqlAnonymizer) miscellaneousNodesProcessor(msg protoreflect.Message) (e
+			// - AT_SetRelOptions, AT_ResetRelOptions, AT_ReplaceRelOptions, AT_GenericOptions
+			// - AT_DropOf, AT_ReAddComment, AT_SetAccessMethod
+		}
+	}
+	return nil
+}
+
+// anonymizeStringNodes walks a slice of *pg_query.Node whose concrete
+// value is expected to be *pg_query.String and replaces every Sval
+// with the deterministic token from the IdentifierHashRegistry.
+//
+// parts may be 1-3 items: [obj] | [schema,obj] | [db,schema,obj].
+// Pass the PREFIX for the *last* element (obj).
+// The helper will automatically use DATABASE_KIND_PREFIX and SCHEMA_KIND_PREFIX
+// for the first and second element when present.
+func (a *SqlAnonymizer) anonymizeStringNodes(nodes []*pg_query.Node, finalPrefix string) error {
+	if len(nodes) == 0 {
+		return nil
+	}
+
+	// pick the actual *pg_query.String values
+	var strs []*pg_query.String
+	for _, n := range nodes {
+		if s := n.GetString_(); s != nil && s.Sval != "" {
+			strs = append(strs, s)
 		}
-		col.Sval, err = a.registry.GetHash(COLUMN_KIND_PREFIX, col.Sval)
+	}
+	if len(strs) == 0 {
+		return nil // nothing to do
+	}
+
+	// build prefix slice based on number of parts
+	var prefixes []string
+	switch len(strs) {
+	case 1:
+		prefixes = []string{finalPrefix} // [obj]
+	case 2:
+		prefixes = []string{SCHEMA_KIND_PREFIX, finalPrefix} // [schema,obj]
+	case 3:
+		prefixes = []string{DATABASE_KIND_PREFIX, SCHEMA_KIND_PREFIX, finalPrefix} // [db,schema,obj]
+	default:
+		return fmt.Errorf("qualified name with %d parts not supported", len(strs))
+	}
+
+	// apply prefixes to each string based on the index
+	for i, s := range strs {
+		tok, err := a.registry.GetHash(prefixes[i], s.Sval)
 		if err != nil {
-			return fmt.Errorf("anon ColumnRef, fieldLen=3, column field: %w", err)
-		}
-	case 4: // db . schema . table . col
-		db := strFields[0]
-		sch := strFields[1]
-		tbl := strFields[2]
-		col := strFields[3]
-		db.Sval, err = a.registry.GetHash(DATABASE_KIND_PREFIX, db.Sval)
+			return err
+		}
+		s.Sval = tok
+	}
+	return nil
+}
+
+// anonymizeColumnRefNode rewrites the String parts of a ColumnRef
+// The slice may have 1–4 parts but the *last* one is always a column.
+func (a *SqlAnonymizer) anonymizeColumnRefNode(strNodes []*pg_query.Node) error {
+	n := len(strNodes)
+	if n == 0 {
+		return nil
+	}
+
+	// 1. Everything *left* of the column (0,1 or 2 items)
+	if n > 1 {
+		// left part could be [tbl] or [sch,tbl] or [db,sch,tbl]
+		if err := a.anonymizeStringNodes(strNodes[:n-1], TABLE_KIND_PREFIX); err != nil {
+			return err
+		}
+	}
+
+	// 2. The right-most item – always a column
+	return a.anonymizeStringNodes(strNodes[n-1:], COLUMN_KIND_PREFIX)
+}
+
+// sample rangevar node: typevar:{catalogname:"db"  schemaname:"schema1"  relname:"mycomposit"  relpersistence:"p"  location:12}
+func (a *SqlAnonymizer) anonymizeRangeVarNode(rv *pg_query.RangeVar, finalPrefix string) (err error) {
+	if rv == nil {
+		return nil
+	}
+	if rv.Catalogname != "" {
```

Comment:
nit: high-level comment for all such conditions, there is already a empty string check in `GetHash` which returns empty string only, so do we really this check everywhere? We can rely on the GetHash to return the empty string in all such cases, and I don't think there will be any problem with that.

Reply:
covered in second PR.